superduper/components/datatype.py

```diff
@@ -168,23 +171,77 @@ def check(data: t.Any) -> bool:
 
     @staticmethod
     @abstractmethod
-    def create(data: t.Any) -> "DataType":
+    def create(data: t.Any) -> "BaseDataType":
         """Create a DataType for the data.
 
         :param data: The data to create the DataType for
         """
         raise NotImplementedError
 
 
-class DataType(Component):
+class BaseDataType(Component):
+    """Base class for datatype.
+
+    :param shape: size of vector
+    """
+
+    type_id: t.ClassVar[str] = 'datatype'
+    # TODO this can just be an integer
+    shape: t.Optional[int] = None
+
+    @abstractmethod
+    def encode_data(self, item, info: t.Optional[t.Dict] = None):
+        """Decode the item as `bytes`.
+
+        :param item: The item to decode.
+        :param info: The optional information dictionary.
+        """
+
+    @abstractmethod
+    def decode_data(self, item, info: t.Optional[t.Dict] = None):
+        """Decode the item from bytes.
+
+        :param item: The item to decode.
+        :param info: The optional information dictionary.
+        """
+
+    def encode_data_with_identifier(self, item, info: t.Optional[t.Dict] = None):
+        b = self.encode_data(item=item, info=info)
+        if isinstance(b, bytes):
+            return b, hashlib.sha1(b).hexdigest()
+        else:
+            return b, hashlib.sha1(str(b).encode()).hexdigest()
+
+
+class NativeVector(BaseDataType):
+    """Datatype for encoding vectors which are supported natively by databackend.
+
+    :param dtype: Datatype of array to encode.
+    """
+
+    dtype: str = 'float64'
+
+    def __post_init__(self, db, artifacts):
+        self.encodable_cls = Native
+        return super().__post_init__(db, artifacts)
+
+    def encode_data(self, item, info=None):
+        if isinstance(item, numpy.ndarray):
+            item = item.tolist()
+        return item
+
+    def decode_data(self, item, info=None):
+        return numpy.array(item).astype(self.dtype)
```

Comment:
`datatype.decode(datatype.encode([1, ,2 ,3])) == np.array([1, 2, 3])` 

The data type has been changed here.

Reply:
I don't understand your point. This is to make sure that, e.g. `numpy.float32` comes back out as `numpy.float32`. If  we don't do this, then it doesn't work, since `x.tolist()` converts to full floating point.

Reply:
Yes agreed. It works in the opposite direction too - if it's a `numpy.array` with `float32` then it also gets changed when converted to list and saved to MongoDB (for example).